### ISSUE TITLE
`NoNameShadowing` false positive in kotlin-dsl

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#5328]
I've been getting `NoNameShadowing` violations in a project that uses `kotlin-dsl` where the name being shadowed is `implicit lambda parameter 'it'`. I think it might be because of the lambda getting converted into an `Action` because there's no other `it` in scope.

```
public fun Project.registerDetektTask(
  name: String,
  sourceSet: KotlinSourceSet
): TaskProvider<Detekt> = tasks.register("detekt${name.capitalize()}", Detekt::class.java) {
  val detekt = detekt
  detekt.basePath?.let { basePath = it } // NoNameShadowing
}
```

**Comment by BraisGabin** [comment]
I assume that the lambda of `register` has a parameter `it` and that's the issue. Can that be the case?

**Comment by eygraber** [comment]
I believe it does, but I think `kotlin-dsl` applies a compiler plugin that converts it into a lambda with receiver. So there is no `it` and the compiler won't let you assign a name to the parameter (because it doesn't exist anymore).

**Comment by BraisGabin** [comment]
I see the problem... But I don't think detekt should take care of this strange behaviour were a compiler plugin is involved. If those functions would annotated with something we could use that to ignore them but that's not the case here.

I think that here you should use a `@Suppress` or write it like this:

```kotlin
  detekt.basePath?.let { b -> basePath = b } // NoNameShadowing
```

**Comment by eygraber** [comment]
I ended up specifying a name to get rid of the violation, but it could be annoying if there was a lot of them.

If detekt used a compiler plugin, would it make it easier to avoid this issue (or would it solve it outright)?

**Comment by BraisGabin** [comment]
I imagine that with a compiler plugin this could be spotted.

Closing for the reasons I wrote before.

**Comment by TWiStErRob** [comment]
+2

I had the same issue with a different rules: just throwing my case here to maybe help the case of getting support for this in some form. I encountered two weird reports, one from `Deprecated` and another from `NestedScopeFunctions`. Only the Deprecated case is detailed here, as it was a bigger WTF. Copied some context from external classes near the code for more visibility:

![image](https://user-images.githubusercontent.com/2906988/194757263-937c9e1c-ab1a-4f95-b23f-0620e362a3ba.png)

**Comment by eygraber** [comment]
Ran into this again with Deprecation and NoNameShadowing in a different projects that had a lot of violations. Was very annoying to address.

**Comment by BraisGabin** [comment]
#5492 should fix this issue. And I understand that this is annoying but we are facing a limitation on our current engine. That's the reason of #5492.

**Comment by eygraber** [comment]
Very excited for that. I'm ready to test 😅
